tools/fixserialnumber.go

```diff
@@ -0,0 +1,188 @@
+			}
+			if len(nextBatch) == 0 || minID >= maxID {
+				close(jobs)
+				close(results)
+				log.Printf("Done. %d/%d errors. %d/%d changed.", errorCount, total, changedCount, total)
+				return
+			}
+			total += len(nextBatch)
+			for _, j := range nextBatch {
+				jobs <- j
+				minID = j.id
+			}
+		}
+	}()
+	for result := range results {
+		if result.err != nil {
+			errorCount++
+			log.Printf("Received error for cert id %d: %s", result.id, result.err)
+		}
+		if result.changed {
+			changedCount++
+		}
+	}
+}
+
+func fetchNextBatchWithRetries(retries int, db *database.DB, minID int64, batchSize int64) (jobs []job, err error) {
+	for i := 0; i < retries; i++ {
+		jobs, err = fetchNextBatch(db, minID, batchSize)
+		if err == nil {
+			break
+		}
+	}
+	return
+}
+
+func fetchNextBatch(db *database.DB, minID int64, batchSize int64) ([]job, error) {
+	rows, err := db.Query(`SELECT id, serial_number, raw_cert
+							   FROM certificates
+							   WHERE id > $1
+							   ORDER BY id
+							   LIMIT $2`,
+		minID,
+		batchSize,
+	)
+	if err != nil {
+		log.Fatalf("Error querying database: %s", err)
+	}
+	defer rows.Close()
+	var jobs []job
+	for rows.Next() {
+		var j job
+		var b64Crt string
+		if err = rows.Scan(&j.id, &j.currentSerialNumber, &b64Crt); err != nil {
+			return nil, fmt.Errorf("Error scanning row: %s", err)
+		}
+		cert, err := b64RawCertToX509Cert(b64Crt)
+		if err != nil {
+			log.Printf("Error converting database certificate to crypto/x509 certificate: %s", err)
+			continue
+		}
+		j.cert = cert
+		jobs = append(jobs, j)
+	}
+	return jobs, nil
+}
+
+func b64RawCertToX509Cert(b64Crt string) (*x509.Certificate, error) {
+	rawCert, err := base64.StdEncoding.DecodeString(b64Crt)
+	if err != nil {
+		return nil, fmt.Errorf("Error b64 decoding certificate: %s", err)
+	}
+	cert, err := x509.ParseCertificate(rawCert)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing x509 certificate: %s", err)
+	}
+	return cert, nil
+}
+
+func worker(id int, jobs <-chan job, results chan result, db *database.DB) {
+	for j := range jobs {
+		correctSerialNumberBytes, err := getSerialNumberASN1Value(j.cert)
+		if err != nil {
+			results <- result{id: j.id, err: err}
+			continue
+		}
+		correctSerialNumber := fmt.Sprintf("%X", correctSerialNumberBytes)
+		if correctSerialNumber == *j.currentSerialNumber {
+			// Serial number is already stored correctly in the database
+			results <- result{id: j.id, err: nil}
+			continue
+		}
+		err = updateSerialNumberInDB(db, j.id, correctSerialNumber)
+		if err != nil {
+			results <- result{
+				id:  j.id,
+				err: fmt.Errorf("Error updating serial number in database: %s", err),
+			}
+			continue
+		}
+		results <- result{id: j.id, err: nil, changed: true}
+	}
+}
+
+// getSerialNumberASN1Value returns the ASN1-encoded serial number for the given certificate, excluding the tag and length.
+func getSerialNumberASN1Value(cert *x509.Certificate) ([]byte, error) {
```

Comment:
I think this is the function we want to move to the certificate package. I'd make it `GetHexSerial` and return the hexadecimal string, since this is what we always manipulate.